Add feature to print unfulfilled requests as a picklists PDF

norrismei opened the pull request:
<!--Read comments, before committing pull request read checklist again

# Checklist:

- I have performed a self-review of my own code,
- I have commented my code, particularly in hard-to-understand areas,
- I have made corresponding changes to the documentation,
- I have added tests that prove my fix is effective or that my feature works,
- New and existing unit tests pass locally with my changes ("bundle exec rake"),
- Title include "WIP" if work is in progress.

-->

Resolves #4406 <!--fill issue number-->

### Description
Stakeholder asked for new functionality that allows printing a single file containing all outstanding requests so that pickers may use the documents as picklists, checking off as they go and making note of any differences.

When an organization goes to their Requests, there will now be a `Print Unfulfilled Picklists` button next to the `Export Requests` button.

Design decisions made:
- Page numbers are based on the number of pages in the PDF file instead of numbering each request page `1 of 1` or `1 of 2` because if somehow the pages were scrambled up, we'd want a way of putting them back in order vs. risking not being able to identify multiple non-descript `2 of 2` pages.
- Format of the Picklists PDF follows the Distribution PDF closely. The `OpenSans` font does not support checkboxes so I chose to go with `[ ]` (open and close brackets) to get the job done with minimal effort.
- Filename of PDF is Picklists_#{Date}_#{Time}.pdf in case the file is generated and printed multiple times in a day, we'd want different file names based on time.

### Things to note
- It takes a little while for the PDF to generate in the testing site because there are 180+ outstanding requests. A query was ran to see how many outstanding requests a real bank would have on average. We're assuming about 10 outstanding requests on average, with the most being 44 requests at the time of the query.

### Type of change

<!-- Please delete options that are not relevant. -->

* New feature (non-breaking change which adds functionality)

### How Has This Been Tested?
I manually tested the following scenarios:
- With packs enabled, the PDF should have the `Units (if applicable)` column. 
To enable, I pointed my browser to http://localhost:3000/flipper and logged in
userid: admin
password: password
- The PDF shows `Partner Pickup Person` if there is one listed under the Partner settings. I logged in as a Partner to toggle this setting.

### Screenshots
No packs enabled and no partner pickup person:

<img width="811" alt="Screen Shot 2024-08-29 at 1 00 09 PM" src="https://github.com/user-attachments/assets/40e76735-e509-4c29-8dd6-8009d288cfb8">

With packs enabled and no partner pickup person:

<img width="808" alt="Screen Shot 2024-08-29 at 12 55 02 PM" src="https://github.com/user-attachments/assets/3cef55c1-9792-4a54-846c-2470accb52d5">

With packs enabled and a partner pickup person listed:
<img width="811" alt="Screen Shot 2024-08-29 at 1 05 07 PM" src="https://github.com/user-attachments/assets/eaf4b9e3-86ea-45f0-bf4d-cdfad340dc3e">

Index page with print button displaying unfulfilled requests count
<img width="1535" alt="Screen Shot 2024-09-23 at 11 29 35 PM" src="https://github.com/user-attachments/assets/fd43a0c9-6f2b-4f74-83dc-622dd949246d">

awwaiid commented:
Ready for a review @norrismei?

norrismei commented:
@awwaiid There are a few changes I need to make and then I'll fill out the PR description and mark this ready for review.

cielf reviewed (commented):
Hey @norrismei!  Good job so far! 

 I am a little concerned about perceived performance (prod is slower than our locals),  but I'll let @awwaiid/@dorner chime in on that from a technical pov.  

(Edit:  I could see adding the number of requests that will be in the picklist into the button.  Just as an indication of what they are getting into.)

cielf commented:
@norrismei  Please do change the button to have the number of unfulfilled requests.. 

e.g.  Print Unfulfilled Picklists (44)

dorner commented on app/controllers/requests_controller.rb:
```diff
@@ -40,6 +40,20 @@ def start
   end
 
+  def print_unfulfilled
+    requests = Request.includes(partner: [:profile]).where(status: [0, 1])
```
Wait - this isn't scoped by organization, no? This will pull every request from every organization.

dorner replied:
This should also include `item_requests` - we should be using those instead of `request_items`, which is a leftover from the days when we had two databases.

dorner replied:
Also, can you use `status: [Request.statuses[:pending], Request.statuses[:started]]`? That's more obvious than 0 and 1.

norrismei replied:
@dorner I'm working to address your 1st and 3rd comment. I have a question about the 2nd one:

> This should also include item_requests - we should be using those instead of request_items, which is a leftover from the days when we had two databases.

I recall talking about `item_requests` and `request_items` in office hours and learning that one was the chosen way going forward but I had it the other way around that `request_items` was what we were keeping. Since I lack the historical context for this, I want to double check before making changes.

dorner replied:
It's `item_requests` that's the real one. That's a fully fledged relational model, as opposed to a serialized JSON object which is harder to work with.

dorner commented on config/routes.rb:
```diff
@@ -241,6 +241,9 @@ def set_up_flipper
     member do
       post :start
     end
+    collection do
+      get :print_unfulfilled
+    end
```
```suggestion
    get :print_unfulfilled, on: :collection
```

dorner reviewed (changes requested):
Great first pass!

norrismei commented:
@cielf I added the number of unfulfilled requests to the print button, and yes, unfulfilled requests includes "started" requests.

@dorner I scoped the query to the organization (thanks for catching) and updated statuses to be more obvious. Awaiting clarity on `item_requests` vs `request_items`.

cielf commented:
Hey @norrismei - just noting that the lint is failing -- just a wee thing to change there.

cielf commented:
@norrismei, but do please change to @item_requests -- we are going to be stripping out request_items entirely in the near future.

dorner commented on app/pdfs/picklists_pdf.rb:
```diff
@@ -0,0 +1,171 @@
+
+        move_down 20
+
+        items = request.item_requests
```
These are basically line items, not items. Probably best to rename.

dorner commented on app/pdfs/picklists_pdf.rb:
```diff
@@ -0,0 +1,171 @@
+      "Differences / Comments"]]
+
+    data + items.map do |i|
+      item_name = Item.find(i.item_id).name
```
This isn't necessary now - you can just do `i.item.name`.

dorner commented on app/controllers/requests_controller.rb:
```diff
@@ -40,6 +40,24 @@ def start
+  def print_unfulfilled
+    requests = current_organization
+      .requests
+      .includes(:item_requests, partner: [:profile])
```
This should be `.includes(item_requests: :item)` to eager-load items as well.

norrismei replied:
I looked at where I used `item_request.item` and saw it was only for the `item.name`. Since there is already `item_request.name`, I updated to use that, which seems it might even be preferable to `item.name` and removed the need to load `item` at all.

dorner commented on app/pdfs/picklists_pdf.rb:
```diff
@@ -0,0 +1,171 @@
+      "Differences / Comments"]]
+
+    data + items.map do |i|
+      item_name = Item.find(i.item_id).name
```
ditto

dorner commented on app/views/requests/index.html.erb:
```diff
@@ -61,6 +61,12 @@
                 <%= clear_filter_button %>
                 <%= modal_button_to("#calculateTotals", {text: "Calculate Product Totals", icon: nil, size: "md", type: "success"}) %>
                 <span class="float-right">
+                  <% if @unfulfilled_requests.exists? %>
```
I usually use `any?` here.

norrismei replied:
Did a little deep dive into `any?` vs `exists?`. Learned that since the records are not pre-loaded, when calling `any?` in the index view, it's delegating to `exists?` anyhow and not loading any records. Then a few lines down on line 67, I'm calling `.count` which runs a second query. I realize for the index view, I only need the count and not the loaded records; therefore, I changed the `@unfulfilled_requests` instance variable to `@unfulfilled_requests_count` in the Requests controller.

dorner reviewed (changes requested):
This needs a unit test please!

norrismei commented:
@dorner unit tests added. Both for the PDF and the button on the index page

cielf reviewed (commented):
LGTM from a functional pov.

cielf commented:
@norrismei  FYI: We had an urgent fix that required all the senior contributors this week , so we didn't get to look at this again.  Hopefully this week will go better.

dorner commented on spec/factories/partners.rb:
```diff
@@ -40,11 +40,20 @@
     end
 
+    transient do
+      pick_up_person { false }
```
Considering this trait is only used once in the entire test suite, it seems like it doesn't need to exist, and the test can create the necessary data itself.

dorner reviewed (dismissed):
One minor comment. Otherwise this looks great!

awwaiid reviewed (approved):
Looks great! Thank you :)
